File: v8-core/src/audit/independence.rs
//! Adversarial Separation & Dual-Key Independence Auditor (D-132, Rule 32).

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DualKeyVerificationResult {
    pub authorized: bool,
    pub implementer_receipt_id: String,
    pub auditor_receipt_id: String,
    pub verdict_hash: String,
}

pub struct IndependenceAuditor;

impl IndependenceAuditor {
    /// Validates dual-key separation of powers: IMPLEMENTER != AUDITOR != VERDICT.
    pub fn audit_dual_key(
        worker_agent_id: &str,
        auditor_agent_id: &str,
        impl_digest: &str,
        audit_replay_digest: &str,
        zero_synthetic_verified: bool,
    ) -> Result<DualKeyVerificationResult, String> {
        // Rule: Worker cannot grade/audit itself
        if worker_agent_id == auditor_agent_id {
            return Err(format!(
                "SELF_GRADING_PROHIBITED: Worker '{worker_agent_id}' cannot act as independent auditor"
            ));
        }

        // Rule: Replay digest must bit-exact match
        if impl_digest != audit_replay_digest {
            return Err(format!(
                "REPLAY_DIGEST_MISMATCH: Impl='{impl_digest}', Audit='{audit_replay_digest}'"
            ));
        }

        // Rule: Zero-synthetic verification must be certified by auditor
        if !zero_synthetic_verified {
            return Err("SYNTHETIC_DATA_LEAKAGE_DETECTED: Auditor failed zero-synthetic check".to_string());
        }

        let mut c = crate::hash::Canon::new();
        c.push_str("DualKeyVerdict");
        c.push_str(worker_agent_id);
        c.push_str(auditor_agent_id);
        c.push_str(impl_digest);
        let verdict_hash = c.finish_blake3_hex();

        Ok(DualKeyVerificationResult {
            authorized: true,
            implementer_receipt_id: format!("impl_{impl_digest}"),
            auditor_receipt_id: format!("audit_{audit_replay_digest}"),
            verdict_hash,
        })
    }
}

File: v8-next/src/v8_next/audit/independence.py
"""Adversarial separation and dual-key independence auditor — port of v8-core/src/audit/independence.rs.

IMPLEMENTER != AUDITOR != VERDICT: the worker may not grade itself, the auditor's replay must
match the implementation digest exactly, and the zero-synthetic check must be certified by the
auditor. The Rust verdict hash uses blake3; Python's standard library has no blake3, so the
digest algorithm is declared as sha256 here and the substitution is stated rather than hidden.
"""

from __future__ import annotations

import hashlib
from dataclasses import dataclass

VERDICT_DIGEST_ALGORITHM = "sha256 (blake3 is not in the Python standard library)"


class IndependenceViolation(Exception):
    """Raised when separation of powers is broken (Rule 32)."""


@dataclass(frozen=True)
class DualKeyVerificationResult:
    authorized: bool
    implementer_receipt_id: str
    auditor_receipt_id: str
    verdict_hash: str


class IndependenceAuditor:
    """Port of ``v8_core::audit::IndependenceAuditor``."""

    @staticmethod
    def audit_dual_key(
        worker_agent_id: str,
        auditor_agent_id: str,
        impl_digest: str,
        audit_replay_digest: str,
        zero_synthetic_verified: bool,
    ) -> DualKeyVerificationResult:
        if worker_agent_id == auditor_agent_id:
            raise IndependenceViolation(
                "SELF_GRADING_PROHIBITED: "
                f"Worker '{worker_agent_id}' cannot act as independent auditor"
            )
        if impl_digest != audit_replay_digest:
            raise IndependenceViolation(
                f"REPLAY_DIGEST_MISMATCH: Impl='{impl_digest}', Audit='{audit_replay_digest}'"
            )
        if not zero_synthetic_verified:
            raise IndependenceViolation(
                "SYNTHETIC_DATA_LEAKAGE_DETECTED: Auditor failed zero-synthetic check"
            )
        payload = "|".join(("DualKeyVerdict", worker_agent_id, auditor_agent_id, impl_digest))
        return DualKeyVerificationResult(
            authorized=True,
            implementer_receipt_id=f"impl_{impl_digest}",
            auditor_receipt_id=f"audit_{audit_replay_digest}",
            verdict_hash=hashlib.sha256(payload.encode()).hexdigest(),
        )
